flink-ml-lib/src/main/java/org/apache/flink/ml/common/broadcast/BroadcastUtils.java

```diff
@@ -22,121 +22,168 @@
                         outType,
-                        env.getParallelism());
-        for (DataStream<?> dataStream : bcStreams.values()) {
+                        parallelism);
+        for (DataStream<?> dataStream : broadcastInputs) {
             transformation.addInput(dataStream.broadcast().getTransformation());
         }
         env.addOperator(transformation);
         return new MultipleConnectedStreams(env).transform(transformation);
     }
 
-    private static String getCoLocationKey(String[] broadcastNames) {
-        StringBuilder sb = new StringBuilder();
-        sb.append("Flink-ML-broadcast-co-location");
-        for (String name : broadcastNames) {
-            sb.append(name);
-        }
-        return sb.toString();
-    }
-
-    private static <OUT> DataStream<OUT> buildGraph(
+    /**
+     * uses {@link DraftExecutionEnvironment} to execute the userDefinedFunction and returns the
+     * resultStream.
+     *
+     * @param env execution environment.
+     * @param inputList non-broadcast input list.
+     * @param broadcastStreamNames names of the broadcast data streams.
+     * @param graphBuilder user-defined logic.
+     * @param <OUT> output type of the result stream.
+     * @return the result stream by applying user-defined logic on the input list.
+     */
+    private static <OUT> DataStream<OUT> getResultStream(
             StreamExecutionEnvironment env,
             List<DataStream<?>> inputList,
             String[] broadcastStreamNames,
             Function<List<DataStream<?>>, DataStream<OUT>> graphBuilder) {
         TypeInformation[] inTypes = new TypeInformation[inputList.size()];
         for (int i = 0; i < inputList.size(); i++) {
-            TypeInformation type = inputList.get(i).getType();
-            inTypes[i] = type;
+            inTypes[i] = inputList.get(i).getType();
         }
-        // blocking all non-broadcast input edges by default.
-        boolean[] isBlocking = new boolean[inTypes.length];
-        Arrays.fill(isBlocking, true);
+        // do not block all non-broadcast input edges by default.
+        boolean[] isBlocked = new boolean[inputList.size()];
         DraftExecutionEnvironment draftEnv =
                 new DraftExecutionEnvironment(
-                        env, new BroadcastWrapper<>(broadcastStreamNames, inTypes, isBlocking));
+                        env, new BroadcastWrapper<>(broadcastStreamNames, inTypes, isBlocked));
 
         List<DataStream<?>> draftSources = new ArrayList<>();
         for (int i = 0; i < inputList.size(); i++) {
             draftSources.add(draftEnv.addDraftSource(inputList.get(i), inputList.get(i).getType()));
         }
         DataStream<OUT> draftOutStream = graphBuilder.apply(draftSources);
-
+        Preconditions.checkState(
+                draftEnv.getStreamGraph(false).getStreamNodes().size() == 1 + inputList.size(),
```

Comment:
Are you sure this would work? When calling `getStreamGraph` the list of `transformations` would be cleared from the env.

Reply:
There is a parameter to decide wheather clear the transformations. Please checkout `StreamExecutionEnvironment#getStreamGraph(boolean clearTransformations)`